hydrator-common/src/main/java/co/cask/hydrator/common/ReferenceBatchSource.java

```diff
@@ -38,5 +44,20 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     super.configurePipeline(pipelineConfigurer);
     // Verify that reference name meets dataset id constraints
     IdUtils.validateId(config.referenceName);
+    // create the external dataset this is done in the configure pipeline stage so that the external dataset with or
+    // without the schema is created when the pipeline is deployed itself.
+    Schema externalSchema = getSchema();
```

Comment:
Instead of using class inheritance for all of this logic, I think it makes more sense to encapsulate the lineage logic in some other class. It would remove code duplication in the source and sink and make it easier to use in other projects. I'm thinking of how we could use this in other projects like the gcp plugins that have their own class hierarchy, where they cannot extend this class or their configs can't extend ReferencePluginConfig. 

I think we can have a LineageRecorder class in hydrator-common that is responsible for creating this dataset if it doesn't exist, and also recording FLL. That way the plugin author doesn't need to know about external datasets and also doesn't have to remember to use the same name for the external dataset as the FLL endpoint.